tokio/src/task/core.rs

```diff
@@ -55,102 +54,213 @@ pub(crate) struct Header {
             header: Header {
                 state,
-                executor: CausalCell::new(None),
-                queue_next: UnsafeCell::new(ptr::null()),
+                queue_next: UnsafeCell::new(None),
                 owned_next: UnsafeCell::new(None),
                 owned_prev: UnsafeCell::new(None),
                 vtable: raw::vtable::<T, S>(),
-                future_causality: CausalCell::new(()),
             },
             core: Core {
-                stage: Stage::Running(Track::new(future)),
+                scheduler: CausalCell::new(None),
+                stage: CausalCell::new(Stage::Running(future)),
             },
             trailer: Trailer {
-                waker: CausalCell::new(MaybeUninit::new(None)),
+                waker: CausalCell::new(None),
             },
         })
     }
 }
 
-impl<T: Future> Core<T> {
-    pub(super) fn transition_to_consumed(&mut self) {
-        self.stage = Stage::Consumed
+impl<T: Future, S: Schedule> Core<T, S> {
+    /// If needed, bind a scheduler to the task.
+    ///
+    /// This only happens on the first poll.
+    pub(super) fn bind_scheduler(&self, task: Task<S>) {
+        use std::mem::ManuallyDrop;
+
+        // TODO: it would be nice to not have to wrap with a ManuallyDrop
+        let task = ManuallyDrop::new(task);
+
+        // This function may be called concurrently, but the __first__ time it
+        // is called, the caller has unique access to this field. All subsequent
+        // concurrent calls will be via the `Waker`, which will "happens after"
+        // the first poll.
+        //
+        // In other words, it is always safe to read the field and it is safe to
+        // write to the field when it is `None`.
+        if self.is_bound() {
+            return;
+        }
+
+        // Bind the task to the scheduler
+        let scheduler = S::bind(&*task);
+
+        // Safety: As `scheduler` is not set, this is the first poll
+        self.scheduler.with_mut(|ptr| unsafe {
+            *ptr = Some(scheduler);
+        });
+    }
+
+    /// Returns true if the task is bound to a scheduler.
+    pub(super) fn is_bound(&self) -> bool {
+        // Safety: never called concurrently w/ a mutation.
+        self.scheduler.with(|ptr| unsafe { (*ptr).is_some() })
     }
 
-    pub(super) fn poll<S>(&mut self, header: &Header) -> Poll<T::Output>
-    where
-        S: Schedule,
-    {
+    /// Poll the future
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure it is safe to mutate the `state field. This
+    /// requires ensuring mutal exclusion between any concurrent thread that
+    /// might modify the future or output field.
+    ///
+    /// The mutual exclusion is implemented by `Harness` and the `Lifecycle`
+    /// component of the task state.
+    ///
+    /// `self` must also be pinned. This is handled by storing the task on the
+    /// heap.
+    pub(super) fn poll(&self, header: &Header) -> Poll<T::Output> {
```

Comment:
It seems from the above that this should be an `unsafe fn`?

Reply:
The `task` module is an unsafe module. It exposes a safe API. As above, module private fns that require the caller to uphold invariants are not annotated as `unsafe`. Doing so would allow arbitrary unsafe usage in the implementation.